mergekit/metric_methods/all_metrics.py

```diff
@@ -0,0 +1,373 @@
+                            model_1_heads.view(model_1_heads.shape[0], -1)],
+                            return_heatmap=True))
+        res.update(SMAPE([model_0_heads.view(model_0_heads.shape[0], -1),
+                            model_1_heads.view(model_1_heads.shape[0], -1)]))
+
+        return res
+
+    def group_label(self) -> Optional[str]:
+        # Use max of the group labels
+        return max([gather_tensor.group_label() for gather_tensor in list(self.weights.values())]) # Check this (X)
+    
+    def __hash__(self):
+        return hash(self.weight_info)
+
+    def __eq__(self, other):
+        if not isinstance(other, AttnTask):
+            return False
+        return self.weight_info == other.weight_info
+
+class DummyTask(Task[torch.Tensor]):
+    gather_tensors: GatherTensors
+    weight_info: WeightInfo
+    def uses_accelerator(self) -> bool:
+        return True
+
+    def arguments(self) -> Dict[str, Task]:
+        return {"tensors": self.gather_tensors}
+
+    def execute(
+        self, **_kwargs
+    ) -> torch.Tensor:
+        
+        return 
+    
+    def group_label(self) -> Optional[str]:
+        return self.gather_tensors.group_label()
+
+# Metric method
+   
+class AllMetric(MetricMethod):
+    attn_weight_tensors: Optional[list] = []
+    attn_weight_infos: Optional[list] = []
+
+    attn_weight_dict: Optional[Dict[str, torch.Tensor]] = {}
+    attn_info_dict: Optional[Dict[str, WeightInfo]] = {}
+
+    attn_parts: Optional[List[str]] = ['k_proj', 'v_proj', 'q_proj', 'o_proj'] # hard-coded for now
+    block_count: Optional[int] = 0
+    def make_task(
+        self,
+        *,
+        output_weight: WeightInfo,
+        parameters: Optional[Dict[str, Any]] = None,
+        tensors: GatherTensors,
+        **_kwargs,
+    ) -> Task:
+        
+        if 'self_attn' in output_weight.name:
+            # collect all attention weights
+            for part in self.attn_parts: # also check only one key
+                if part in output_weight.name:
+                   self.attn_weight_dict[part] = tensors
+                   self.attn_info_dict[part] = output_weight   
+
+            # if all attention weights are collected, create attention task
+            if set(list(self.attn_weight_dict.keys())) == set(self.attn_parts):
+                weights, infos = self.attn_weight_dict, self.attn_info_dict
+                self.attn_weight_dict, self.attn_info_dict = {}, {}
+                weight_info = WeightInfo(
+                    name=f"Attention Block {self.block_count}",
+                    force_dtype=None,
+                    optional=False,
+                    aliases=None,
+                    GQA_groups=4, # hard-coded for now
+                    num_heads=32 # hard-coded for now
+                )
+                self.block_count += 1
+                return AttnTask(weights=weights, weight_infos=infos, weight_info=weight_info)
```

Comment:
Does this end up creating N AttnTasks for each block? I don't think it's actually a problem as the tasks will be deduplicated downstream - should be fine

Reply:
Should only be one AttnTask for each block - the if statement on line 351 is only satisfied once all the tensors (K,Q,V,O) have been collected. Then  self.attn_weight_dict is reset to {} and the (one) AttnTask is created. I might also add individual tensor metrics for comparing just the Qs, Vs etc, which would be simpler.